packages/astro/src/config.ts

```diff
@@ -1,108 +1,105 @@
-      throw new Error(`[config] ${key}: ${JSON.stringify(config[key])}\n  Expected string, received ${type(config[key])}.`);
-    }
-  }
-
-  // booleans
-  for (const key of ['sitemap']) {
-    if (config[key] !== undefined && config[key] !== null && typeof config[key] !== 'boolean') {
-      throw new Error(`[config] ${key}: ${JSON.stringify(config[key])}\n  Expected boolean, received ${type(config[key])}.`);
-    }
-  }
-
-  // buildOptions
-  if (config.buildOptions) {
-    // buildOptions.site
-    if (config.buildOptions.site !== undefined) {
-      if (typeof config.buildOptions.site !== 'string') throw new Error(`[config] buildOptions.site is not a string`);
-      try {
-        new URL(config.buildOptions.site);
-      } catch (err) {
-        throw new Error('[config] buildOptions.site must be a valid URL');
-      }
-    }
-  }
-
-  // devOptions
-  if (typeof config.devOptions?.port !== 'number') {
-    throw new Error(`[config] devOptions.port: Expected number, received ${type(config.devOptions?.port)}`);
-  }
-  if (typeof config.devOptions?.hostname !== 'string') {
-    throw new Error(`[config] devOptions.hostname: Expected string, received ${type(config.devOptions?.hostname)}`);
-  }
-  if (config.devOptions?.tailwindConfig !== undefined && typeof config.devOptions?.tailwindConfig !== 'string') {
-    throw new Error(`[config] devOptions.tailwindConfig: Expected string, received ${type(config.devOptions?.tailwindConfig)}`);
-  }
-}
-
-/** Set default config values */
-async function configDefaults(userConfig?: any): Promise<any> {
-  const config: any = { ...(userConfig || {}) };
-
-  if (config.projectRoot === undefined) config.projectRoot = '.';
-  if (config.src === undefined) config.src = './src';
-  if (config.pages === undefined) config.pages = './src/pages';
-  if (config.dist === undefined) config.dist = './dist';
-  if (config.public === undefined) config.public = './public';
-  if (config.devOptions === undefined) config.devOptions = {};
-  if (config.devOptions.port === undefined) config.devOptions.port = await getPort({ port: getPort.makeRange(3000, 3050) });
-  if (config.devOptions.hostname === undefined) config.devOptions.hostname = 'localhost';
-  if (config.devOptions.trailingSlash === undefined) config.devOptions.trailingSlash = 'ignore';
-  if (config.buildOptions === undefined) config.buildOptions = {};
-  if (config.buildOptions.pageDirectoryUrl === undefined) config.buildOptions.pageDirectoryUrl = true;
-  if (config.markdownOptions === undefined) config.markdownOptions = {};
-  if (config.buildOptions.sitemap === undefined) config.buildOptions.sitemap = true;
-
-  return config;
-}
+import path from 'path';
+import { z } from 'zod';
+import { AstroConfig, AstroUserConfig } from './@types/astro';
+
+export const AstroConfigSchema = z.object({
+  projectRoot: z
+    .string()
+    .optional()
+    .transform((val) => new URL(val + '/')),
```

Comment:
What if the provided value has a trailing slash?

Reply:
I’m not sure! I just copied this from existing code so not a regression, but I agree we should have an “addTrailingSlash” helper similar to snowpack for this sort of thing

Reply:
Works for me!

Reply:
done!